### ISSUE TITLE
Regression: width() only works when explicitly set

### ISSUE THREAD

**Issue by Krinkle** [jquery/jquery#3571]
### Description ###

Select any element on the page that doesn't have an explicit width set via a stylesheet or inline attribute and call `.width()`.

In jQuery 1.x, 3.0.x and 3.1.1 this returns the computed width.

As of jQuery 3.2.0 it returns `0`.

### Link to test case ###

<http://codepen.io/Krinkle/pen/RpjgpR>

**Comment by mgol** [comment]
Ugh, it's hard to believe we didn't have a single unit test for that...

I've created a milestone for 3.2.1, this seems serious enough that we'd want a quick patch release before 3.3.0.

**Comment by mgol** [comment]
Caused by #3561.

**Comment by mgol** [comment]
Not having style specified via stylesheet or inline is not enough to trigger the issue; the element has to have a computed `display: inline` value.

The underlying issue is that `getComputedStyle` on inline elements returns `auto`.

**Comment by mgol** [comment]
From the spec `getComputedStyle(elem)` returns an object with resolved values for various CSS properties for `elem`. The definition of the resolved value is at:
https://www.w3.org/TR/cssom-1/#resolved-value
For `width` resolved value is defined in the following way:

> If the property applies to the element or pseudo-element and the resolved value of the display property is not none, the resolved value is the used value. Otherwise the resolved value is the computed value.

The `width` property doesn't apply to inline elements so the computed value is returned instead of the used one (which it doesn't have). And the computed value will always be `auto` (again, as the property doesn't apply to inline elements).

Whether `width` is set inline or in the stylesheet on a particular element doesn't matter. The issue is that `.width()` is broken in 3.2.0 for all inline elements.

**Comment by mgol** [comment]
I'm not sure what we can do to fix this issue and still keep #3193 (i.e. not take transforms into account while computing width/height) & #1724 (support for subpixel values for width/height) fixed. This issue proves `getComputedStyle` is not enough but `offsetWidth` doesn't support fractional values while `getBoundingClientRect()` takes transforms into account.

Choose your poison. :/

**Comment by Krinkle** [comment]
@mgol Yeah, the problem is that the semantic purpose behind `.width()` is split between two very different concepts. On the one side the use case "computed style", with the rough expectation that setting the same value in CSS would essentially be a no-op (ignoring border-box for now). This doesn't apply to inline elements since setting a width does nothing.
```lang=html
<span>foo</span>
<span style="width: 900px;">foo</span>
```
These will have the same effective width, unless inline-block or block is applied to the span. So it is true that inline elements don't have a width other than 'auto'.

On the other side, there is the use case of "visual size", which semantically maps to `getBoundingClientRect()` and is often used to decide where to position an element, or to decide whether text will fit in an element. (The first test failure I found when attempting to upgrade to jQuery 3.2.0 was a test failure from an old plugin called [jquery.autoEllipsis](https://github.com/wikimedia/mediawiki/blob/49748181dd56ec97e7ba7c13e684a16abceb3cc0/resources/src/jquery/jquery.autoEllipsis.js))

The problem is that until transforms came along, there wasn't really any way the visual size could be different from the size behind the scenes. The missing concept here is "internally drawn size".

However, I would not surprise me if some users of jQuery `width()` actually expect transforms to be taken into account. It is not obvious from historical record and documentation what the expectation is, and regardless, these are two separate use cases. We cannot satisfy both with the same method :-/.

**Comment by mgol** [comment]
@Krinkle you're right `.width()` has too many responsibilities; this discussion already partially happened in #3193. The problem is any real solution for that would require decoupling those responsibilities which creates a breaking change.

As for what we can do in 3.2.1, anything we do to fix this issue will break something a previous version promised to fix; at this point I think we should just revert Timmy's PR as that will break the thing that we fixed most recently so few people will be relying on that at this point.

The real solution will have to wait for 4.0.

**Comment by craigkovatch** [comment]
> As for transforms, I'm not convinced that it's obvious that that they shouldn't be taken into account.

I agree, but I wonder if a caller-specified parameter (i.e. shouldIncludeTransforms or some such) would be the easier and more flexible path?

**Comment by HolgerJeromin** [comment]
> I agree, but I wonder if a caller-specified parameter

The whole api is designed as read: .width() and write: .width(newValue)
So adding a parameter is not easy possible.
I think we really need a new api for proper support the whole mess: .width(), .css('width')

**Comment by timmywil** [comment]
@craigkovatch @HolgerJeromin Right, but even if we added another method (or signature or whatever) to support adding transforms, there's still this issue of how to support getting width with no transforms correctly.

I wonder if falling back to `offsetWidth` is the best option. I still lean towards ignoring transforms and erring on the side of whole numbers rather fractional values, but only when the value we get from `curCSS` is `auto`.

**Comment by craigkovatch** [comment]
@timmywil I agree. Better to be missing a fraction of a pixel than be unexpectedly transformed by some scaling factor.

**Comment by timmywil** [comment]
I'm swamped at work right now. Can someone pick this up today?

**Comment by timmywil** [comment]
I'll do it tonight.

**Comment by mohithg** [comment]
My app relies on width() and height() on inline-elements and updating to 3.2 broke my app too.

**Comment by mgol** [comment]
@mohithg The fixed version is available now: https://blog.jquery.com/2017/03/20/jquery-3-2-1-now-available/

**Comment by mohithg** [comment]
@mgol Thanks.

**Comment by Krinkle** [comment]
The fix didn't work for Android 4. Could re-open, but filed #3602 instead.

**Comment by Camzilla** [comment]
@mgol surprisingly ran into this problem today when replacing a table with flex. Items in my rows have "flex: 0 0 value%" and are showing up as '0' width for .height() .innerHeight() and also .outerHeight(). I got jQuery 3.2.1 in my project, is flex accounted for?

**Comment by mgol** [comment]
@Camzilla This seems unrelated to this issue. Could you open a new one with a test case?
